---
title: "Class_exercise"
author: "Byanka Estudillo"
date: "3/16/2023"
output: github_document
---


### Load the libraries you need
### Load functions you need "my_class_functions"
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)

source("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/util/my_class_functions.R")

load("results/peak_features.RData", verbose = T)
```


### load in your peak files for each replicate of each protein
### Here I am starting to analyze my data for my proteins of interest:
### protein CEBPZ, CHD2, CTCF, ELF1, and EP300
### First I will read in each replicate file
```{r load in peak files}
# filepath to import peaks
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/byes1293"
peak_path <- "group/results/bwa/mergedLibrary/macs/broadPeak/"
broadpeakfilepath <- file.path(basepath, peak_path)

# printing out a table of the number of peaks in each file:
# import peaks
peak_list <- import_peaks(consensus_file_path = broadpeakfilepath)
# let's get a list of how many peaks are in each file before we create consensus peaks.
peak_num <- sapply(peak_list, length) %>% as.data.frame(row.names = T)
# label column
names(peak_num) <- c("num_peaks")

# make dbp name a col.
peak_num <- peak_num %>%
  rownames_to_column(var = "dbp") %>%
  separate(col = dbp,  into = c('dbp', 'replicate'), sep = "_")

peak_num
```


### Now I am going to create consensus peaks for each protein
```{r consensus peaks}
# first we need a list of unique dbps as input parameter
dbps <- unique(sapply(names(peak_list), function(x) {
   unlist(strsplit(x, "_"))[1]
}))

# now run our function consensus_from_reduced
consensus_list <- lapply(dbps, consensus_from_reduced, peak_list)
names(consensus_list) <- dbps

dbps

# let's see the num_peaks for consensus 
num_consensus_peaks <- sapply(consensus_list, length) %>% 
  as.data.frame() %>%
  rownames_to_column( var = "dbp") %>%
  dplyr::rename(number_consensus_peaks = ".")

# merge into num_peaks
peak_num <- left_join(peak_num, num_consensus_peaks)

peak_num

# export consensus peaks to results folder
# setting file path to export
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/byes1293"
consensus_path <- "CLASS_2023/CLASSES/05_R_analyses/class_exercise1/results/consensus_peaks/"
exportpath <- file.path(basepath, consensus_path)

# now we can export each as .bed file
for(i in 1:length(consensus_list)) {
rtracklayer::export(consensus_list[[i]], paste0(exportpath, names(consensus_list)[i], "_consensus_peaks.bed") )}


```

### Now I am going to make my consensus peaks compatable with UCSC genome browser
```{r}
# FIRST a file list:
consensus_file_list <- list.files("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/CLASSES/05_R_analyses/class_exercise1/results/consensus_peaks", full.names = T, pattern = ".bed")

# lapply (for loop) across consensus file list to add colnames
# The actual col names for .broadPeak are: chr, start, end, name, score, strand
peaks <- lapply(consensus_file_list, read.table, col.names = c("chr", "start", "end", "name", "score", "strand"))

names(peaks) <- dbps

# make chromosomes of interest object
canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")

# let's use lapply with filter funciton to cannonical_chr
peaks <- lapply(peaks, function(x) x %>% filter(chr %in% canonical_chr))


# now that these are all nice and clean let's export:
new_filenames <- paste0("results/consensus_peaks/", names(peaks), "_consensus.bed")

for(i in 1:length(peaks)) {
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}

# we are using paste0 to print the header text
# then we are adding the names as the value the header equals with 'names' function
headers <- paste0("track type=bed name=", names(peaks))
headers

# print out consensus peak files in a results/UCSC directory
# creating a path to export after we add header in for loop below
new_filenames <- paste0("results/ucsc_consensus_peaks/", names(peaks), ".bed")
new_filenames

# let's do so more indexing and only print out the first two of the three files.

for(i in 1:length(peaks)) {
  # Write the header line
  writeLines(headers[[i]], new_filenames[[i]])
  # Append the broadPeak table data
  
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}
```


### Now I want to compare a protein with a previous analysis 
```{r}

# goto UCSC genome browser and load in a peak file for a given protein
# load in the data for the same protein from the previous analysis
# compare how your consensus peaks are similar or different to previous analyses
knitr::include_graphics("ucsc_screenshot.png")

#My consensus peaks for ELF1 are similar to the previous analyses. The peaks line up really well.

```


### Now I am going to determine how my peaks for each protein overlap annotations of the genome
### First I will find the overlaps between my consensus peaks with promoters of lncRNA and mRNA promoters

```{r}
# starting annotation file that we will use moving forward.
lncrna_mrna_genes <- rtracklayer::import("results/gene_annotations/mrna_lncrna_genes.gtf")

# Nice that loaded so much faster -- lets see as DF
lncrna_mrna_genes_df <- lncrna_mrna_genes %>% as.data.frame()

# importing promoter annotations
lncrna_mrna_promoters <- rtracklayer::import("results/gene_annotations/lncrna_mrna_promoters.gtf")

# last handy annotation will be lncRNA and mRNA gene IDs to subset
lncrna_gene_ids <- mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "lncRNA"]
table(mrna_lncrna_genes$gene_type)
# same for mRNAs
mrna_gene_ids <-mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "protein_coding"]


# find overlaps of promoters for each protein
# filepath to import peaks
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/byes1293"
peak_path <- "CLASS_2023/CLASSES/05_R_analyses/class_exercise1/results/consensus_peaks"
consensusPeakPath <- file.path(basepath, peak_path)

consensus_peaks_files <- list.files(consensusPeakPath, 
                                             pattern = "*consensus.bed",
                                             full.names = TRUE)

# lapply with import function to make a list of GRanges
consensus_peaks <- lapply(consensus_peaks_files, rtracklayer::import)

# adding DBP name to each list of GRanges from the file name
# note gsub requires no space around or argument ( | )
names(consensus_peaks) <- gsub("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/CLASSES/05_R_analyses/class_exercise1/results/consensus_peaks/|_consensus.bed","", consensus_peaks_files)

# Let's start with loading in the number of peaks each DBP has -- using length.
num_peaks_df <- data.frame("dbp" = names(consensus_peaks),
                           "num_peaks" = sapply(consensus_peaks, length))


# Now let's get the total amount of the genome covered by all the peaks for a given DBP.
num_peaks_df$total_peak_length <- sapply(consensus_peaks, function(x) sum(width(x)))

# counting promoter overlaps
promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, consensus_peaks, type = "counts")

# ok we see this is a large matrix of the number of overlaps at each promoter
# cols are promoters, rows are DBPS
# so now we just row sum for each DBP !
num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)
```

## results: 
#1) What can you determine from these overlaps?
CEBPZ has more peaks overlapping promoters than its total number of peaks. CHD2 has almost the same numer of peaks overlapping promoters as its total number of peaks. It looks like those two proteins especially tend to bind near promoters.


### Now I want to compare the overlaps with lncRNA and mRNA promoters seperately 
```{r}
# Now let's break these promoters into two groups "lncrna" and "mrna"
# We will use the gene_id objects we made above to index and separate them.
num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])

# mrna promoter overlaps
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])
```
## results:
# 1) What is the difference in overlaps between mRNA and lncRNA promoters
Looks like the number of peaks overlapping lncrna promoters tend to be a lot less than the number of peaks overlapping mrna promoters. EP300 seems to be an exception where the number of peaks overlapping lncrna promoters is more than the number of peaks overlapping mrna promoters.


### Now I am going to test if there is more binding over gene bodies than promoters
### I will seperate lncRNA and mRNA gene bodies to find the overlaps 

```{r}
# Finding overlaps with gene_bodies (will take a few minutes again)
# Note this takes 10-15 min
genebody_peak_counts <- count_peaks_per_feature(mrna_lncrna_genes, 
                                                consensus_peaks, 
                                                type = "counts")

# Now let's extract the overlaps the same way we did for promoters above

# All gene bodies
num_peaks_df$peaks_overlapping_genebody <- 
  rowSums(genebody_peak_counts)

# lncRNA gene bodies 
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])

# mRNA gene bodies
num_peaks_df$peaks_overlapping_mrna_genebody <- 
  rowSums(genebody_peak_counts[,mrna_gene_ids])

write_csv(num_peaks_df, "results/num_peaks_df.csv")
```
## results: 
# 1) Do my proteins have more overlaps with promoters or genebodies?
Overall, it looks like my proteins overlap more with genebodies than promoters.


### I am curious if my proteins are transcription factors so I will use the annotations
### in a cell paper I found and see

```{r}
url <- "https://www.cell.com/cms/10.1016/j.cell.2018.01.029/attachment/ede37821-fd6f-41b7-9a0e-9d5410855ae6/mmc2.xlsx"

destination_for_url <- "results/TF_annotations.xlsx"

# to download we can use download.file
download.file(url, destination_for_url)

#redx1::read_excel to import
human_tfs <- readxl::read_excel("results/TF_annotations.xlsx",
                                sheet = 2, skip = 1)


# let's rename the 4th column to indicate if it is a TF.
names(human_tfs)[4] <- "is_tf"

# now let's intersect gene names that are in our ChIP data and has TF identity.
length(which(tolower(num_peaks_df$dbp) %in% tolower(human_tfs$Name)))

# first let's filter and grab the first 4 columns that match DBPs in num_peaks_df
human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(num_peaks_df$dbp), 1:4]


# adding new column names
names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

# Now comes the actual merge. Let's look at it first
# ?merge

num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)

# Let's check how many NAs -- we should have some missing values.
dim(num_peaks_df[is.na(num_peaks_df$tf),])

num_peaks_df <- num_peaks_df[,1:12]

num_peaks_df
```


### It is nice and all to find overlaps, but I am interested in how many proteins
### bind a specific promoter. I will use my handy "occurence" parameter in 
### " count peaks per feature" 

```{r}
# note similar as above but using type = occurnece
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, consensus_peaks, 
                                               type = "occurrence")

# Let's double check that all lncrna & mrna genes are accounted for:
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))

# First make sure promoter_peak_occurrence and lncrna_mrna_promoters are in the same order
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))


# Now let's use the 'data.frame()' fucntion. Set up a bunch of colnames and populate them.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "1kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))

```
## results: I find the max number of proteins on a promoter to be 4


### Now I want to start plotting my results
### First I will see if there is a relationship between peak number and total DNA covered
```{r}
source("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/util/plotting_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/util/_setup.R")

# let's make this simple plot first: number of peaks -vs- total peak length
ggplot(num_peaks_df, aes(x = num_peaks, 
                         y = total_peak_length)) +
  geom_point(shape = 'triangle',
             color = 'dark green') 
```

### Now I want to color my plot by wether the protein is a TF or not.
```{r}
ggplot(num_peaks_df, aes(x = num_peaks, 
                 y = total_peak_length,
                 color = tf == "Yes")) +
  geom_point(shape = 'triangle')

```

### I want to make a histogram of the number of peaks for each of my proteins

```{r}
ggplot(num_peaks_df, aes(x = num_peaks, fill = tf)) +
  geom_histogram(bins = 10)

```


### Now I want to facet this by the type of DNA binding domain my protein has.
```{r}
ggplot(num_peaks_df, aes(x = num_peaks, fill = tf)) +
  facet_wrap(dbd ~ .) +
  geom_histogram(bins = 10)
```


### Cool now I am ready to send my result to my collaborator as a
### Knitted document

