---
title: "Final_class_project"
author: "Byanka Estudillo"
date: "05/01/2023"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(tidyverse)
library(ggpubr)
library(ggplot2)
source("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/util/my_class_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/util/_setup.R")
```

# Your mission should you choose to accept it 
There is an enormous amount of data in ENCODE that 
needs to be analyzed. This includes 480 unique DBP
ChIPseq datasets. To facilitate your mission we provide you
1,066 peak files that have been run through the NF_CORE chipseq
pipeline. 

You will find all the .broadPeak peak files here:
/scratch/Shares/rinnclass/CLASS_2023/data/data/peaks

Use what you have learned in class to uncover
new results from this daunting data set. 
# Good luck the genome depends on you!

# first you might want to think about your organization
# 1) create a directory in your class folder for these analyses
    # -> folder name
          -> util/
            -> scripts.R
          -> analysis/
            -> 01_consensus_peaks
              -> .RMD
              -> results/
              -> figures
            -> 02_clustering etc....
  
# Consensus peaks
# Num_peaks_df
# Peak_occurence_df
# lncRNA and or mRNA promoters / gene annotations
# SAVE THIS !! save(object_name, nextobject, ...., file = "where you save this.Rdata")
# load(file.Rdata)


# creating filtered consensus peaks
```{r}
# filepath to import peaks
broadpeakfilepath <- "/scratch/Shares/rinnclass/CLASS_2023/data/data/peaks"

# using import peaks to import .broadPeak files
peak_list <- import_peaks(consensus_file_path = broadpeakfilepath)

# Creating unique DBP object for create_consensus_peaks_from_reduced
dbp <- unique(sapply(names(peak_list), function(x) {
   unlist(strsplit(x, "_"))[1]
}))

# now run our function consensus_from_reduced
consensus_list <- lapply(dbp, consensus_from_reduced, peak_list)

# adding names to the GRange list
names(consensus_list) <- dbp

### Filter out any chip data less 1,000 peaks == filtered consensus peaks ###
# filtering to 1000 peaks
filtered_consensus_list <- consensus_list[sapply(consensus_list, length) > 1000]

# setting file path to export
exportpath <- "/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/CLASSES/Final_Rmd/results/filtered_consensus_peaks/"

for(i in 1:length(filtered_consensus_list)) {
  rtracklayer::export(filtered_consensus_list[[i]], 
                      paste0(exportpath, 
                             names(filtered_consensus_list)[i], 
                             "_filtered_consensus_peaks.bed"))
}
```


# loading in genome features
```{r}
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/gencode.v32.annotation.gtf")

# gencode genes
gencode_genes <- gencode_gr[gencode_gr$type == "gene"] 

# mrna_genes
mrna_genes <- gencode_genes[gencode_genes$gene_type %in% "protein_coding"]

# lncrna_genes
lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% "lncRNA"] 

# mrna_lncrna_genes
mrna_lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% c("protein_coding","lncRNA")]

# lncrna_mrna_promoters
lncrna_mrna_promoters <- promoters(mrna_lncrna_genes, upstream = 1000, downstream = 1000)

# lncrna_gene_ids
lncrna_gene_ids <- mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "lncRNA"]

# mrna_gene_ids
mrna_gene_ids <-mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "protein_coding"]
```


# making data frame of filtered_consensus_peak info
```{r}
num_peaks_df <- data.frame("dbp" = names(filtered_consensus_list),
                           "num_peaks" = sapply(filtered_consensus_list, length))

# total genome covered by peaks
num_peaks_df$total_peak_length <- sapply(filtered_consensus_list, function(x) sum(width(x)))

# creating number of promoter overlaps entry
promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_list, type = "counts")

# creating promoter peak_occurence for clustering - Metaplots later.
promoter_peak_matrix <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_list, type = "occurrence")

# saving
# write.table(promoter_peak_matrix, "results/promoter_peak_occurrence_matrix.tsv")

# read back in
promoter_peak_occurrence_matrix <- read.table("results/promoter_peak_occurrence_matrix.tsv")

# summing rows to get total number of promoter overlaps
num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)

# lncrna promoter overlaps 
num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])

# mrna promoter overlaps
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])

# Finding overlaps with gene_bodies (will take a few minutes again)
# Note this takes several minutes
genebody_peak_counts <- count_peaks_per_feature(mrna_lncrna_genes, 
                                                filtered_consensus_list, 
                                                type = "counts")

# All gene bodies overlaps
num_peaks_df$peaks_overlapping_genebody <- rowSums(genebody_peak_counts)

# lncRNA gene bodies 
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])

# mRNA gene bodies
num_peaks_df$peaks_overlapping_mrna_genebody <- rowSums(genebody_peak_counts[,mrna_gene_ids])
```


# creating promoter peak occurence matrix
```{r}
# running count_peaks_per_feature
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_list, 
                                               type = "occurrence")

# Let's double check that all lncrna & mrna genes are accounted for:
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))

# Now let's use the 'data.frame()' fucntion. Set up a bunch of colnames and populate them.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "1kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))
```


# plotting num_peaks versus genome coverage.
```{r}
### How does peak number and genome coverage compare ###
ggplot(num_peaks_df, aes(x = num_peaks, y = total_peak_length)) +
  geom_point() + 
  geom_smooth(method = "gam", se = TRUE, color = "black", lty = 2)+
  ylab("BP covered") +
  xlab("Number of peaks") +
  ggtitle("Peak count vs. total bases covered")

# saving 
# ggsave("figures/peak_num_vs_coverage.pdf")
```
Result: there is a linear relationship between number of peaks and total coverage


# Promoter overlaps vs gene bodies
```{r}
### What is the distribution of promoter overlaps versus gene-bodies (hint hist) ###
## Histogram of peaks overlapping with promoters
hist(num_peaks_df$peaks_overlapping_promoters)

# Histogram of peaks overlapping gene bodies
hist(num_peaks_df$peaks_overlapping_genebody)
```

# Finding super binders
```{r}
### Make a list of genes that are "super binders" ###
ggplot(peak_occurence_df, aes(x = number_of_dbp)) +
geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes") +
  facet_wrap(~gene_type)

#ggsave("figures/super_binder_lncRNA_mRNA_curves.pdf", height = 5, width = 8)

# super binding promoters have greater than 200 DBPs bound
# add indicator for super binders
peak_occurence_df <- peak_occurence_df %>% 
  mutate(super_binder = ifelse(number_of_dbp < 200, "not super binder", "super binder"))
```


# Comparing mRNA and lncRNA promoter overlaps
```{r}
### Is there a difference in mRNA and lncRNA promoter overlaps? ###
ggplot(num_peaks_df, aes(x = num_peaks)) +
  geom_point(aes(y = peaks_overlapping_lncrna_promoters), color = "red") +
  geom_point(aes(y = peaks_overlapping_mrna_promoters), color = "black") +
  #stat_regline_equation(aes(y = peaks_overlapping_lncrna_promoters), color = "red") +
  #stat_regline_equation(aes(y = peaks_overlapping_mrna_promoters), color = "black", label.y = 20000) +
  geom_smooth(aes(y = peaks_overlapping_lncrna_promoters), method = "lm", se = FALSE, formula = "y ~ x") +
  geom_smooth(aes(y = peaks_overlapping_mrna_promoters), method = "lm", se = FALSE, formula = "y ~ x")

# saving
#ggsave("figures/lncRNA-mRNA-promoter_binding.pdf")
```
Results: mRNA promoters have more binding than lncRNA promoters.


```{r}
### Do lncRNAs also have super-binding promoters? ###
# Subsetting list of super binding lncRNA promoters 
lncrna_super_binders <- subset(peak_occurence_df, gene_type == "lncRNA" & super_binder == "super binder")

nrow(lncrna_super_binders)

# Subsetting list of super binding mRNA promoters 
mrna_super_binders <- subset(peak_occurence_df, gene_type == "protein_coding" & super_binder == "super binder")

nrow(mrna_super_binders)
```
Results: I found 2521 lncRNA superbinding promoters and 9201 mRNA super-promoters


# Adding TF type annotations for DBPs
```{r}
human_tfs <- readxl::read_excel("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/CLASSES/Final_Rmd/results/TF_annotations.xlsx", skip = 1)

names(human_tfs)[4] <- "is_tf"

length(which(tolower(num_peaks_df$dbp) %in% tolower(human_tfs$Name)))

human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(num_peaks_df$dbp), 1:4]

names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)

#write_csv(num_peaks_df, "results/num_peaks_df.csv")

### How many of these proteins are TFs? What is the most represented type of DBD? ###
yes_tf <- subset(num_peaks_df, tf == "Yes")
nrow(yes_tf)
no_tf <- subset(num_peaks_df, tf == "No")
nrow(no_tf)
```
Result: Exactly 344 of our DBPs are TFs and 63 of our DBPs are not TFs. That leaves 25 of our DBPs as neither. 


# saving key env objects
```{r}
#save(filtered_consensus_list, gencode_genes, lncrna_gene_ids, lncrna_mrna_promoters, mrna_gene_ids, num_peaks_df, peak_occurence_df, promoter_peak_occurrence_matrix, lncrna_mrna_promoters, mrna_lncrna_genes, file = "results/peak_features.RData")

#load("results/peak_features.RData", verbose = T)
```

# Let's see how similar our genes are to each other
# Clustering
```{r}
promoter_peak_occurence_matrix <- read.table("analysis/01_create_consensus_peaks/results/promoter_peak_occurrence_matrix.tsv")

# creating distance matrix
peak_occurence_dist <- dist(promoter_peak_occurence_matrix, method = "binary")

# clustering distance values
bin_hier <- hclust(peak_occurence_dist, method = "complete")

# Dendrogram of binding profiles by promoter (not binding profile - below)
ggdendro::ggdendrogram(bin_hier, rotate = FALSE,  size = 3, 
                       theme_dendro = TRUE) +
   coord_flip() +
   scale_y_continuous() +
   scale_x_continuous(position = "top") +
   scale_x_continuous(breaks = seq_along(bin_hier$labels[bin_hier$order]),
             labels = bin_hier$labels[bin_hier$order], position = "top",
             expand = c(0,0)) +
   theme(axis.text.x = element_text(angle = 90, hjust  = 1)) + 
   theme(axis.text.y = element_text(angle = 0,hjust = 1)) +
   scale_y_reverse(expand = c(0.01, 0)) +
   theme(
     plot.background = element_blank(),
     panel.grid.major = element_blank(),
   panel.grid.minor = element_blank(),
     panel.border = element_blank()
   )
#ggsave("figures/promoter_overlap_dendogram.pdf", height = 50, width = 12, limitsize = F)

### if we cluster by lncRNA and mRNA separately what are some similarities and differences? ###
lncrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "lncRNA"]

mrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "protein_coding"]

# for lncRNAs
lncrna_peak_occurence <- promoter_peak_occurence_matrix[,lncrna_promoters$gene_id]

length(lncrna_peak_occurence)

bin_hier_lncrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))

ggdendro::ggdendrogram(bin_hier_lncrna, rotate = T,  size = 3)
#ggsave("figures/lncrna_hclust_binary_dist.pdf", height = 49, width = 6)

# for mRNAs
mrna_peak_occurence <- promoter_peak_occurence_matrix[,mrna_promoters$gene_id]

bin_hier_mrna <- hclust(dist(mrna_peak_occurence, method = "binary"))

ggdendro::ggdendrogram(bin_hier_mrna, rotate = T,  size = 3)
#ggsave("figures/mrna_hclust_binary_dist.pdf", height = 44, width = 6)
```


# saving key objects
```{r}
#save(promoter_peak_occurence_matrix, peak_occurence_dist, bin_hier, lncrna_peak_occurence, lncrna_promoters, bin_hier_lncrna, mrna_peak_occurence, mrna_promoters, bin_hier_mrna, file = "results/peak_features.RData")

#load("results/clusters.RData", verbose = T)
```


# Metaplots
```{r}
### Let's look at the metaplot for all DBPs on lncRNA and mRNA promoters separately (hint facet wrap) ###
#setting up lncrna DF.
lncrna_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# for loop to populate DF with overlap density in lncrna promoters
suppressWarnings({
  for(i in 1:length(filtered_consensus_list)) {
  print(names(filtered_consensus_list)[[i]])
  tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters = lncrna_promoters)
  tmp_df$dbp <- names(filtered_consensus_list)[[i]]
  lncrna_metaplot_df <- bind_rows(lncrna_metaplot_df, tmp_df)
  
  }
})

# saving
#write_rds(lncrna_metaplot_df, "results/lncRNA_metaplot_df_final.rds")

# now for mRNAs 
mrna_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# for loop to populate mRNA_metaplot
suppressWarnings({
  for(i in 1:length(filtered_consensus_list)) {
  print(names(filtered_consensus_list)[[i]])
  tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters = mrna_promoters)
  tmp_df$dbp <- names(filtered_consensus_list)[[i]]
  mrna_metaplot_df <- bind_rows(mrna_metaplot_df, tmp_df)
  
  }
})

# saving mRNA metaplots
#write_rds(mrna_metaplot_df, "results/mrna_metaplot_df_final.rds")

# now adding the information of gene type
mrna_metaplot_df$gene_type <- "mRNA"
lncrna_metaplot_df$gene_type <- "lncRNA"
combined_metaplot_profile <- bind_rows(mrna_metaplot_df, lncrna_metaplot_df)

# saving
#write_rds(mrna_metaplot_df, "results/metaplot_df_final.rds")

# plot
ggplot(combined_metaplot_profile,
  aes(x = x, y = dens, color = gene_type )) +
  geom_vline(xintercept = 0, lty = 2) +
  geom_line(size = 1.5) +
  facet_wrap(dbp ~ ., scales = "free_y") +
  ggtitle("Promoter Metaplot") +
  scale_x_continuous(breaks = c(-1000, 0, 1000),
                  labels = c("-1kb", "TSS", "+1kb"),
                  name = "") +
  ylab("Peak frequency") +
  scale_color_manual(values = c("#424242","#A8404C"))
#ggsave("figures/lncrna_mrna_metaplot_profile.pdf", width = 40, height = 10)
```


# superbinder vs nonsuper binder metaplots
```{r}
### Make a metaplot of DBPS only on Super-binders versus regular promoters ... ###
# What promoters make up these super binders
super_binders <- peak_occurence_df %>% 
  filter(peak_occurence_df$number_of_dbp > 200)

#write_csv(super_binders, "results/super_binders.csv")

# What promoters have less than 200 dbps bound
non_super_binders <- peak_occurence_df %>% 
  filter(peak_occurence_df$number_of_dbp < 200)

#write_csv(non_super_binders, "results/non_super_binders.csv")

# subset mRNA and lncRNA promoters by super binders
super_gr <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_id %in% super_binders$gene_id]
non_super_gr <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_id %in% non_super_binders$gene_id]

#setting up lncrna df
super_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# for loop to populate DF with overlap density in lncrna promoters
suppressWarnings({
  for(i in 1:length(filtered_consensus_list)) {
  print(names(filtered_consensus_list)[[i]])
  tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters = super_gr)
  tmp_df$dbp <- names(filtered_consensus_list)[[i]]
  super_metaplot_df <- bind_rows(super_metaplot_df, tmp_df)
  
  }
})

# saving
#write_rds(super_metaplot_df, "results/super_metaplot_df_final.rds")

# now for mRNAs 
non_super_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())

# for loop to populate mRNA_metaplot
suppressWarnings({
  for(i in 1:length(filtered_consensus_list)) {
  print(names(filtered_consensus_list)[[i]])
  tmp_df <- profile_tss(filtered_consensus_list[[i]], lncrna_mrna_promoters = non_super_gr)
  tmp_df$dbp <- names(filtered_consensus_list)[[i]]
  non_super_metaplot_df <- bind_rows(non_super_metaplot_df, tmp_df)
  
  }
})

# saving mRNA metaplots
#write_rds(non_super_metaplot_df, "results/non_super_metaplot_df_final.rds")

# now adding the information of gene type
super_metaplot_df$gene_type <- "super"
non_super_metaplot_df$gene_type <- "not super"
combined_metaplot_profile <- bind_rows(super_metaplot_df, non_super_metaplot_df)

ggplot(combined_metaplot_profile, 
       aes(x = x, y = dens, color = gene_type )) +
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  facet_wrap(dbp ~ ., scales = "free_y") +
  ggtitle("Super Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-1000, 0, 1000),
                     labels = c("-1kb", "TSS", "+1kb"),
                     name = "") + 
  ylab("Peak frequency") +
 scale_color_manual(values = c("#424242","#a8404c"))

# saving
#ggsave("figures/super_mega_meta_plot.pdf", width = 49, height = 12)
```


# saving key objects
```{r}
#save(lncrna_metaplot_df, mrna_metaplot_df, combined_metaplot_profile, super_binders, super_metaplot_df, non_super_binders, non_super_metaplot_df, file = "results/metaplot_data.RData")

#load("results/metaplot_data.RData", verbose = T)
```


# RNAseq expression
```{r}
# First let's read in the sample sheet to know what is what
samplesheet <- read_rds("/scratch/Shares/rinnclass/CLASS_2023/byes1293/CLASS_2023/CLASSES/05_R_analyses/05_RNAseq/01_differential_expression/results/final_samplesheet.rds")

# reading in salmon tpm
salmon_tpm <- read.csv("/scratch/Shares/rinnclass/CLASS_2023/data/data/rnaseq/results/salmon/salmon_merged_gene_tpm.csv")

# TPM table is in same order as samplesheet
tpm <- salmon_tpm %>% 
  pivot_longer(cols = 2:ncol(.), names_to = "sample_id", values_to = "tpm") %>%
  merge(samplesheet) %>%
  group_by(gene_id, condition) %>%
  summarize(tpm = mean(tpm, na.rm = T)) %>%
  pivot_wider(names_from = condition, values_from = tpm, names_prefix = "tpm_")

# Now we can merge in the TPM data to this data.frame
# This will merge on the on the gene_id column
promoter_features_df <- merge(peak_occurence_df, tpm)

# saving this file
#write.csv(promoter_features_df, "results/promoter_feature_df_tpm.csv")

### What is the relationship between number of DBPS bound on a promoter versus RNA output (hint TPM) ###
# plotting binding vs total RNA expression
ggplot(promoter_features_df, 
            aes(y = log2(tpm_homo_sapiens_hepg2 + 0.001), x = number_of_dbp, color = gene_type)) + 
geom_point(data = promoter_features_df %>% filter(tpm_homo_sapiens_hepg2 > 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  #stat_cor() +
  geom_smooth(method = "lm") +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 

#ggsave("figures/binding_vs_expression_total_rna.pdf")
```
Result: DBPs bound to a promoter are positively correlated with RNA expression. Protein coding genes have higher expression than lncRNA genes. Expression of protein coding genes seem to have a more linear relationship with number of DBPs bound compared to lncRNA genes.


# Binding versus nuclear expression
```{r}
# Now let's make a similar plot for nuclear RNA abundance versus #DBPs bound to their promoter
ggplot(promoter_features_df, 
            aes(y = log2(tpm_homo_sapiens_nuclear_fraction + 0.001), x = number_of_dbp, color = gene_type)) + 
  geom_point(data = promoter_features_df %>% filter(tpm_homo_sapiens_nuclear_fraction > 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  #stat_cor() +
  
  
  #making look nice
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Nuclear Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 

# saving figure
#ggsave("figures/nuclear_expression-vs-promoter_binding.pdf")
```
Result: looks very similar to total RNA binding versus expression


# Binding versus cytoplasmic expression
```{r}
# Same thing just seeing if there is a difference of cyto RNAs versus DBPs on promoter
ggplot(promoter_features_df, 
            aes(y = log2(tpm_homo_sapiens_cytosolic_fraction + 0.001), x = number_of_dbp, color = gene_type)) + 
  geom_point(data = promoter_features_df %>% filter(tpm_homo_sapiens_cytosolic_fraction > 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  #stat_cor() +
  
  
  # making look nice
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Cytoplasmic Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 
# saving figure
#ggsave("figures/cytoplasmic_expression-vs-promoter_binding.pdf")
```
RESULTS:
(1) everything seems to be low abundance 
(2) Some mRNAs are expressed in the nucleus -- we could look at this more later.
(3) The same linear trend holds but is driven by mostly low expression events.


# lncRNA vs mRNA expression in total RNA
```{r}
# plotting
ggplot(promoter_features_df, aes(x = log2(tpm_homo_sapiens_hepg2 + 0.01), color = gene_type))+
  geom_density()

# saving figure
#ggsave("figures/mrna_lncrna_tpm_total_rna.pdf")

# let's also do the same for nuclear since lncRNAs are typically more nuclear
ggplot(promoter_features_df, aes(x = log2(tpm_homo_sapiens_nuclear_fraction + 0.01), color = gene_type))+
  geom_density()

# saving figure
#ggsave("figures/mrna_lncrna_tpm_nuclear.pdf")
```
RESULT:
(1) This yet again confirms lncRNAs have lower expression levels than mRNAs.
(2) In the nuclear fraction it shift's to closer.
(3) lot's of mRNA with nuclear expression -- that seems odd


# super binding vs expression
```{r}
ggplot(promoter_features_df, 
            aes(y = log2(tpm_homo_sapiens_hepg2 + 0.001), x = number_of_dbp, color = super_binder)) + 
geom_point(data = promoter_features_df %>% filter(tpm_homo_sapiens_hepg2 > 0.001),
             shape = 17, alpha = 0.7) +
  
  # Adding a generative additive model 
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  # this adds the statistics from the gam to the figure
  #stat_cor() +
  geom_smooth(method = "lm") +
  # this is just to make it look nice.
  scale_x_continuous(expand = c(0,0)) +
  # adding colors manually
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Super Binder") + 
  # title
  ggtitle("Expression vs. promoter binding events (split by super binder)") + 
  # labeling axes
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 

#ggsave("figures/binding_vs_expression_total_rna_super.pdf")
```
Result - If we look at the trend for only the super promoters, it is linear but less steep than the trend for the lower binding promoters. There are still a large number of super binders with low expression. 


# Zooming in
```{r}
### If we zoom in on high binding promoters (> 200 DBPs) are there any that don't have any expression? ###
zoomed_df <- promoter_features_df[promoter_features_df$super_binder == "super binder",]

ggplot(zoomed_df, 
            aes(y = log2(tpm_homo_sapiens_hepg2 + 0.001), x = number_of_dbp, color = gene_type)) + 
geom_point(data = zoomed_df %>% filter(tpm_homo_sapiens_hepg2 > 0.001),
             shape = 17, alpha = 0.7) +
  
  # Adding a generative additive model 
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  # this adds the statistics from the gam to the figure
  stat_cor() +
  geom_smooth(method = "lm") +
  # this is just to make it look nice.
  scale_x_continuous(expand = c(0,0)) +
  # adding colors manually
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  # title
  ggtitle("Expression vs. promoter binding events (zoomed in on super binders)") + 
  # labeling axes
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 

#ggsave("figures/binding_vs_expression_total_rna_zoomed.pdf")
```
Result - mRNA super binders have higher expression levels, but also pretty large variance in expression levels, and many of the cases with low expression are in mRNA super binders. The trend between binding and expression is very similar between lncRNA and mRNA super binders.


# Now with nuclear expression
```{r}
# Now let's make a similar plot for nuclear RNA abundance versus #DBPs bound to their promoter
ggplot(promoter_features_df, 
            aes(y = log2(tpm_homo_sapiens_nuclear_fraction + 0.001), x = number_of_dbp, color = super_binder)) + 
  geom_point(data = promoter_features_df %>% filter(tpm_homo_sapiens_nuclear_fraction > 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Super binder") + 
  ggtitle("Nuclear Expression vs. promoter binding events (separated by super binder)") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 
  
# saving figure
#ggsave("figures/nuclear_expression-vs-promoter_binding_super.pdf")
```
Result: The result for nuclear expression looks very similar to the overall result.


# Binding vs cytoplasmic expression
```{r}
# Same thing just seeing if there is a difference of cyto RNAs versus DBPs on promoter
ggplot(promoter_features_df, 
            aes(y = log2(tpm_homo_sapiens_cytosolic_fraction + 0.001), x = number_of_dbp, color = super_binder)) + 
            # note if we take off the log2 everything is very lowly expressed
  geom_point(data = promoter_features_df %>% filter(tpm_homo_sapiens_cytosolic_fraction > 0.001),
             shape = 17, alpha = 0.7) +
  geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  stat_cor() +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#a8404c", "#424242"), name = "Gene type") + 
  ggtitle("Cytoplasmic Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[2](TPM))) 


#ggsave("figures/cytoplasmic_expression-vs-promoter_binding_super.pdf")
```
The trends are similar for super binders vs. not super binders, and not as strong. There are overall many cases where there is low expression with high expression outliers in both super binders and non super binders. 

